pkg/util/region_range_lock.go

```diff
@@ -0,0 +1,337 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package util
+
+import (
+	"bytes"
+	"encoding/hex"
+	"fmt"
+	"math"
+	"sync"
+	"sync/atomic"
+
+	"github.com/google/btree"
+	"github.com/pingcap/log"
+	"github.com/pingcap/tidb/store/tikv/oracle"
+	"go.uber.org/zap"
+)
+
+type rangeTsEntry struct {
+	// Only startKey is necessary. End key can be inferred by the next item since the map always keeps a continuous
+	// range.
+	startKey []byte
+	ts       uint64
+}
+
+func rangeTsEntryWithKey(key []byte) *rangeTsEntry {
+	return &rangeTsEntry{
+		startKey: key,
+	}
+}
+
+func (e *rangeTsEntry) Less(than btree.Item) bool {
+	return bytes.Compare(e.startKey, than.(*rangeTsEntry).startKey) < 0
+}
+
+// RangeTsMap represents a map from key range to a timestamp. It supports range set and calculating min value among a
+// a specified range.
+type RangeTsMap struct {
+	m *btree.BTree
+}
+
+// NewRangeTsMap creates a RangeTsMap.
+func NewRangeTsMap() *RangeTsMap {
+	return &RangeTsMap{
+		m: btree.New(16),
+	}
+}
+
+// Set sets the corresponding ts of the given range to the specified value.
+func (m *RangeTsMap) Set(startKey, endKey []byte, ts uint64) {
+	if m.m.Get(rangeTsEntryWithKey(endKey)) == nil {
+		// To calculate the minimal ts, the default value is math.MaxUint64
+		tailTs := uint64(math.MaxUint64)
+		m.m.DescendLessOrEqual(rangeTsEntryWithKey(endKey), func(i btree.Item) bool {
+			tailTs = i.(*rangeTsEntry).ts
+			return false
+		})
+		m.m.ReplaceOrInsert(&rangeTsEntry{
+			startKey: endKey,
+			ts:       tailTs,
+		})
+	}
+
+	entriesToDelete := make([]*rangeTsEntry, 0)
+	m.m.AscendRange(rangeTsEntryWithKey(startKey), rangeTsEntryWithKey(endKey), func(i btree.Item) bool {
+		entriesToDelete = append(entriesToDelete, i.(*rangeTsEntry))
+		return true
+	})
+
+	for _, e := range entriesToDelete {
+		m.m.Delete(e)
+	}
+
+	m.m.ReplaceOrInsert(&rangeTsEntry{
+		startKey: startKey,
+		ts:       ts,
+	})
+}
+
+// GetMin gets the min ts value among the given range. endKey must be greater than startKey.
+func (m *RangeTsMap) GetMin(startKey, endKey []byte) uint64 {
+	var ts uint64 = math.MaxUint64
+	m.m.DescendLessOrEqual(rangeTsEntryWithKey(startKey), func(i btree.Item) bool {
+		ts = i.(*rangeTsEntry).ts
+		return false
+	})
+	m.m.AscendRange(rangeTsEntryWithKey(startKey), rangeTsEntryWithKey(endKey), func(i btree.Item) bool {
+		thisTs := i.(*rangeTsEntry).ts
+		if ts > thisTs {
+			ts = thisTs
+		}
+		return true
+	})
+	return ts
+}
+
+func (m *RangeTsMap) String() string {
+	result := ""
+	m.m.Ascend(func(i btree.Item) bool {
+		e := i.(*rangeTsEntry)
+		result += fmt.Sprintf("%v %v %v\n", e.ts, oracle.GetTimeFromTS(e.ts), hex.EncodeToString(e.startKey))
```

Comment:
It may print a very large string, is it expected?

Reply:
Oh, it is very long, and I use this to debug the region merge's logic. Maybe I should remove it?

Reply:
If it is only used for debug, then LGTM.